src/DataAPI.jl

```diff
@@ -89,6 +91,9 @@ If the collection is not sortable then the order of levels is unspecified.
 
 Contrary to [`unique`](@ref), this function may return values which do not
 actually occur in the data, and does not preserve their order of appearance in `x`.
+
+If a type of `A` implements `levels` with a meaningful order then it should also
+implement `isordered` as by defualt `isordered` returns `false`.
```

Comment:
This is a good point. I was thinking how to put it best and I just "kind of" reused what CategoricalArrays.jl defined:

> Test whether entries in `A` can be compared using `<`, `>` and similar operators, using the ordering of levels.

Which implies that `isordered` should tell us if what `levels(A)` returns is ordered if we use `<` for comparison of these elements (so technically if `issorted(levels(A))` returns `true`).

In particular I understood that if `levels` returns something for a type that supports `<` that fails `issorted(levels(A))` then `isordered` should return `false`.

This was my understanding of the original intent of what `isordered` should mean i.e. that two conditions must be met:
1. eltype of `A` supports `<` etc.
2. `levels` returns values in order consistent with it

(but maybe I was wrong).

But if I was right then `levels` and `isordered` must be in sync. This is different from `Base.Ordered()`, as `Base.Ordered()` is unrelated with `levels`.

Reply:
Yeah, the problem is that `isordered` was initially designed with only CategoricalArrays in mind, and for them it happens to be the case that `isordered`, `levels` and `<`/`isless` are all consistent. So now to extend it we have to decide whether that should always be the case (and document it) or not.  That would be the simplest and most obvious definition. But it's also quite strict as e.g. a custom array for which `levels(x) == ["c", "b", "a"]` wouldn't respect it. In short, I think we should know what we want to use this trait for to choose the most useful solution.

@quinnj You mentioned at https://github.com/JuliaData/DataAPI.jl/issues/23#issuecomment-705114877 that Arrow.jl could use this for a custom dict-encoded type. How would that type behave? How would it implement `isordered`?

Reply:
The arrow spec says:

> /// By default, dictionaries are not ordered, or the order does not have
  /// semantic meaning. In some statistical, applications, dictionary-encoding
  /// is used to represent ordered categorical data, and we provide a way to
  /// preserve that metadata here
  isOrdered: bool;

So it's not really doing much by itself other than preserving the metadata. The `DictEncoded` array type in Arrow.jl is basically just a `PooledArray`, but also includes the boolean `isOrdered` field; but any further "semantics" would need to be handled by the user or custom comparison code (or in the example of converting to a CategoricalArray, it could query the `DataAPI.isordered` property).

Reply:
OK, thanks. Looking at the [code](https://github.com/JuliaData/Arrow.jl/blob/b109c4921908346c63b1e5a09bb49845aad598b8/src/arraytypes/dictencoding.jl), it seems that you don't override `levels(::DictEncoded)` to get the levels in their custom order, right? So there's no way for the user to access that metadata? IMO it would be problematic to define `isordered(::DictEncoded)` if the custom ordering isn't reflected at all in either `levels` or `<`/`isless` called on the values.

Reply:
@nalimilan - so does it affect this PR?

I still think that the recommendation:

> If a type of `A` implements `levels` with a meaningful order then it should also implement `isordered` as by default `isordered` returns `false`.

makes sense. Note that type of `A` is a concrete type (not `UnionAll`)

Reply:
I would prefer to have a decision here which way we go so that we can merge this (or what we decide is OK), make a release and update https://github.com/JuliaData/DataFrames.jl/pull/2494 accordingly before we merge it.

Reply:
Personally, I don't think we should require `eltype of A supports < etc.` and `levels returns values in order consistent with it`, if we don't have to. The `Arrow.DictEncoded` type won't be able to support that, for example. It includes a `DictEncoding` type, which is its "pool" type, which has values stored as a plain array, and a boolean flag `isOrdered` to indicate whether the order of the pool values is significant or not. But they very well could be in a different order than `sort` would make via `<` or `isless` comparison. And I think that's an important feature of these kind of "factor"/categorical arrays.

But then again, if that totally rocks the boat and we already assume those 2 points in a lot of places, then I'm fine if we commit to that. It just won't fit the Arrow.jl definition, which probably won't be that big of a deal, since people will just have ot check the flat manually and code around the themselves if they have a use-case.

Reply:
I'm doing a round of cleanup on Arrow.jl right now, with implementing DataAPI.jl for `DictEncoded` properly as one of hte TODO items.

Reply:
I would be OK with both. The original definition came from CategoricalArrays.jl which are special, so we can be more flexible here I think (in which case `isordered` is giving us roughly the same as `Ordered` trait in Julia Base).

Reply:
If `Arrow.DictEncoded` doesn't want to expose the custom order of levels using generic API, then I don't think it makes sense for it to implement `isordered`. Otherwise that function will be useless as it won't allow writing generic code that can do something with levels: you would know there's an order, but you wouldn't be able to retrieve it.

One issue with `Base.OrderStyle` is that it's supposed to work when passing only the type, but for CategoricalArray it's a runtime property...

Reply:
But `Arrow.DictEncoded` _does_ want to expose the custom order of levels? I'd like to implement `DataAPI.refarray`, `DataAPI.refpool`, `DataAPI.levels` and `DataAPI.isordered` on DictEncoded, so it can participate in all the optimizations built around those.

Reply:
OK, I thought you didn't want that as you said:

> Personally, I don't think we should require `eltype` of A supports `<` etc. and `levels` returns values in order consistent with it, if we don't have to.

Then we can require `isordered` to be consistent with `levels` at least, but mention that it may be inconsistent with `<` and `isless`. But we shouldn't say that `isordered` should return `true` for types that implement `levels`, as it wouldn't always be the case for `CategoricalArray` and `Arrow.DictEncoded`.

Reply:
Ok, yeah, that makes sense to me. Yeah, for `Arrow.DictEncoded`, `isordered` is an independent property.